tekton/resources/cd/folder-template.yaml

```diff
@@ -75,18 +75,42 @@ spec:
             #!/bin/sh
             set -ex
 
-            # Determine whether to deploy a normal folder or a kustomize overlay
-            METHOD_PARAM="-f"
-            [[ "$(params.isOverlay)" == "true" ]] && METHOD_PARAM="-k"
-
             # Determine whether to enforce namespace across resources
             NAMESPACE_PARAM="-n $(params.namespace)"
             [[ "$(params.namespace)" == "" ]] && NAMESPACE_PARAM=""
 
-            # Run the deployment
-            kubectl "$(params.deployMethod)" \
-              $NAMESPACE_PARAM $METHOD_PARAM \
-              $(resources.inputs.source.path)/$(params.folderPath)
+            # Handle overlays
+            TARGET=$(resources.inputs.source.path)/$(params.folderPath)
+            if [[ "$(params.isOverlay)" == "true" ]]; then
+              TARGET=target.yaml
+              kustomize build \
+                $(resources.inputs.source.path)/$(params.folderPath) > $TARGET
+            fi
+
+            # Check if there is any diff
+            DIFF=diff.txt
+            kubectl diff $NAMESPACE_PARAM -f $TARGET | tee $DIFF
+
+            # If there is no diff, we don't need to update
+            if [ ! -s ${DIFF?} ]; then
+              echo "No change detected, nothing to be done."
+              exit 0
+            fi
+
+            # When deploying with replace, we need to do a create first,
+            # to ensure new resources are created
+            CREATE_OUTPUT=create.txt
```

Comment:
Do we have a way to detect resources that were created before (by this), and not created anymore (aka we should remove them) ?

Reply:
I added a "managed-by" annotation for now, but we could make that a label. That way we could search for all resources managed by tekton and delete those that don't show up in the manifests we're deploying.
We could even add more granularity as we have multiple cronjobs doing deployment, so each should care about its own resources only.

Reply:
can we add support for deletion as a follow-up?

Reply:
yes :+1: